## Introdução 

Este desafio tem como objetivo avaliar sua aptidão para atuar na jornada de dados da LH Nautical, testando sua capacidade analítica, domínio técnico, visão de negócio e pragmatismo. Buscamos entender seu raciocínio de ponta a ponta: desde o tratamento de dados brutos e modelagem SQL até a geração de insights estratégicos e aplicação de lógica preditiva.
Não avaliamos apenas o código, mas a sua habilidade de transformar dados desorganizados em soluções que gerem valor real para a operação. Responda a todas as etapas com base no contexto fictício da LH Nautical descrito abaixo, seguindo as orientações do Tech Lead, Gabriel Santos.


## Contexto Fictício
### Sobre a Empresa

A LH Nautical é uma empresa de varejo náutico com lojas físicas, armazéns e canal de e-commerce. Seus dados operacionais cobrem o período de 2020 a 2026 e registram o ciclo completo: catálogo de produtos, pedidos de venda, pagamentos, nota fiscal eletrônica, compras de fornecedores, movimentação de estoque e devoluções.
Você terá acesso a 24 arquivos CSV na pasta lh_nautical_csv/, representando o schema relacional da empresa. Sua missão é percorrer as etapas de um pipeline de dados real, da ingestão bruta à inteligência aplicada.

Stakeholders Principais
Durante o desafio, você interagirá com as necessidades de três perfis centrais da empresa: 
• Gabriel Santos (Tech Lead): O mentor técnico que valoriza a organização, a documentação e a clareza do raciocínio acima de códigos complexos. 
• Marina Costa (Gerente de Negócios): Focada em resultados práticos, margens de lucro e performance de vendas. 
• Sr. Almir (Fundador): Representa a visão old school; ele desconfia da "nuvem" e precisa ser convencido por dados sólidos e análises precisas. 
Seu Objetivo

Você acaba de receber um e-mail do Gabriel Santos, Tech Lead da LH Nautical. A empresa vai lançar uma campanha na semana que vem, mas os dados estão completamente desorganizados. Sua missão é limpar as bases, estruturar as tabelas e gerar os relatórios que a diretoria exige. Mas o Gabriel avisou: "Eu valorizo mais a organização e a explicação do que o código rodando sem eu entender nada."

Sua missão é atuar como o profissional de dados que transformará esse cenário. Você terá acesso a bases brutas (como o catálogo de produtos e históricos de vendas) e deverá realizar desde a limpeza e modelagem (Engenharia de Dados e SQL) até a geração de insights preditivos e sistemas de recomendação (Ciência de Dados e IA).


## Cenário

Antes de qualquer análise, modelagem ou tomada de decisão, é fundamental entender o que existe nos dados. O Sr. Almir quer uma resposta simples: “Posso confiar nesses dados para tomar decisões?”
Sua missão é realizar uma análise exploratória inicial nas tabelas e responder perguntas básicas, porém críticas, sobre volume, distribuição e qualidade dos dados.

### Premissas obrigatórias

- Utilize apenas a tabela "orders"
- Não faça limpeza nem tratamento dos dados
- Apenas observe, agregue e descreva


In [17]:
## instalação do pandas para os cálculos
%pip install pandas

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


### Tarefas: 

## 1. Visão geral da tabela `orders`

A análise exploratória inicial foi iniciada pela tabela `orders`, pois ela concentra as informações de pedidos e vendas utilizadas nas etapas seguintes do desafio.

Nesta seção, foram realizadas as seguintes verificações:

- carregamento do arquivo `orders.csv` com a biblioteca `pandas`;
- identificação da quantidade de linhas e colunas;
- verificação do período disponível na base, a partir das colunas de data;
- análise dos tipos de dados;
- identificação de valores ausentes e registros duplicados;
- análise descritiva das variáveis numéricas, com foco no valor total dos pedidos;
- investigação de possíveis valores atípicos por meio da metodologia IQR.

Não foi realizado tratamento, remoção ou alteração dos dados. O objetivo desta etapa é compreender a estrutura, a cobertura temporal e os principais pontos de atenção da base antes das análises posteriores.

In [5]:
import pandas as pd


# Carregamento da tabela usada na análise exploratória
orders = pd.read_csv("orders.csv")

Inicialmente, foi verificado o volume da tabela, sua estrutura e o período temporal disponível. Essa etapa permite identificar a cobertura dos dados e o intervalo de negócio analisado.

In [6]:
print(f"Quantidade de linhas: {orders.shape[0]:,}")
print(f"Quantidade de colunas: {orders.shape[1]}")

created_at = pd.to_datetime(orders["created_at"])

print(f"Data mínima: {created_at.min()}")
print(f"Data máxima: {created_at.max()}")

orders.info()

Quantidade de linhas: 48,998
Quantidade de colunas: 13
Data mínima: 2020-01-01 01:19:28
Data máxima: 2026-12-31 23:43:09
<class 'pandas.DataFrame'>
RangeIndex: 48998 entries, 0 to 48997
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               48998 non-null  int64  
 1   order_number     48998 non-null  str    
 2   channel          48998 non-null  str    
 3   customer_id      48998 non-null  int64  
 4   salesperson_id   24867 non-null  float64
 5   location_id      48998 non-null  int64  
 6   status           48998 non-null  str    
 7   subtotal         48998 non-null  float64
 8   discount_amount  48998 non-null  float64
 9   total            48998 non-null  float64
 10  placed_at        48998 non-null  str    
 11  created_at       48998 non-null  str    
 12  updated_at       48998 non-null  str    
dtypes: float64(4), int64(3), str(6)
memory usage: 4.9 MB


## 2. Análise da coluna `total`

Em seguida, foram analisados os valores dos pedidos. Média, mediana, quartis e extremos permitem compreender a distribuição e identificar possíveis pedidos atípicos.

Verificando duplicidade nos identificadores. 
Isso é relevante para falar de possíveis inconsistências na tabela. Quero comparar o número com os 48.998 registros encontrados.

In [7]:
resumo_total = orders["total"].describe()

print(resumo_total)

print(f"\nValor mínimo: R$ {orders['total'].min():,.2f}")
print(f"Valor máximo: R$ {orders['total'].max():,.2f}")
print(f"Valor médio: R$ {orders['total'].mean():,.2f}")

count     48998.000000
mean      28704.992077
std       19425.636818
min          32.620000
25%       13171.235000
50%       25917.840000
75%       40941.882500
max      127262.020000
Name: total, dtype: float64

Valor mínimo: R$ 32.62
Valor máximo: R$ 127,262.02
Valor médio: R$ 28,704.99


### Valor mínimo: R$ 32.62
### Valor máximo: R$ 127,262.02
### Valor médio: R$ 28,704.99

## 3. Interpretação

Responda de forma resumida:
Com base na análise exploratória realizada, escreva um breve diagnóstico sobre a confiabilidade da tabela o para análises futuras.
Comente sobre:
- possíveis outliers em "total",
- qualidade dos dados (valores nulos ou inconsistentes),
- e se você considera que o dataset está pronto para análises ou se exigiria tratamento prévio


### Verificação de valores ausentes e duplicados

Nesta etapa, foi avaliada a qualidade inicial da tabela `orders` por meio da identificação de valores ausentes em cada coluna.

Para cada coluna, foi calculado:

- **Quantidade de valores nulos:** número de registros sem informação.
- **Percentual de valores nulos:** proporção de registros sem informação em relação ao total de linhas.

A fórmula utilizada foi:

\[
\text{Percentual de nulos} =
\frac{\text{Quantidade de valores nulos na coluna}}
{\text{Quantidade total de linhas}}
\times 100
\]

No código, `isnull()` identifica valores ausentes. Como valores booleanos podem ser interpretados como 1 para `True` e 0 para `False`, a função `mean()` calcula diretamente a proporção de nulos na coluna. Por fim, os resultados foram ordenados de forma decrescente para priorizar as colunas com maior ausência de informação.

A análise permite identificar campos que exigem atenção em etapas posteriores, sem realizar qualquer tratamento ou remoção de dados nesta fase.

In [9]:
nulos = orders.isnull().sum().sort_values(ascending=False)
percentual_nulos = (orders.isnull().mean() * 100).sort_values(ascending=False)

qualidade = pd.DataFrame({
    "valores_nulos": nulos,
    "percentual_nulo": percentual_nulos.round(2)
})

qualidade

,valores_nulos,percentual_nulo
salesperson_id,24131,49.25
id,0,0.00
order_number,0,0.00
channel,0,0.00
customer_id,0,0.00
location_id,0,0.00
status,0,0.00
subtotal,0,0.00
discount_amount,0,0.00
total,0,0.00


A ausência de `salesperson_id` pode ser esperada em vendas de e-commerce, mas deve ser validada antes de análises de desempenho de vendedores.

In [12]:
# Verificação de duplicidade em registros
orders[["id", "order_number"]].nunique()

id              48998
order_number    48998
dtype: int64

Os dois retornaram com o mesmo valor: 48.998. Então não foi encontrado duplicidade nos registros.

## 3.1 Análise de outliers pelo método IQR

Para aprofundar a análise da coluna `total`, foi utilizado o método do Intervalo Interquartil (IQR).

O IQR corresponde à diferença entre o terceiro quartil (Q3) e o primeiro quartil (Q1). Foram considerados outliers os valores abaixo de `Q1 - 1,5 × IQR` ou acima de `Q3 + 1,5 × IQR`.

In [19]:
## aplicação da fórmula de IQR 
q1 = orders["total"].quantile(0.25)
q3 = orders["total"].quantile(0.75)

iqr = q3 - q1

limite_inferior = q1 - 1.5 * iqr
limite_superior = q3 + 1.5 * iqr

outliers = orders[
    (orders["total"] < limite_inferior) |
    (orders["total"] > limite_superior)
]

quantidade_outliers = len(outliers)
percentual_outliers = (quantidade_outliers / len(orders)) * 100

print(f"Q1: R$ {q1:,.2f}")
print(f"Q3: R$ {q3:,.2f}")
print(f"IQR: R$ {iqr:,.2f}")
print(f"Limite inferior: R$ {limite_inferior:,.2f}")
print(f"Limite superior: R$ {limite_superior:,.2f}")
print(f"\nQuantidade de outliers: {quantidade_outliers:,}")
print(f"Percentual da base: {percentual_outliers:.2f}%")
print(f"Menor outlier: R$ {outliers['total'].min():,.2f}")
print(f"Maior outlier: R$ {outliers['total'].max():,.2f}")

Q1: R$ 13,171.24
Q3: R$ 40,941.88
IQR: R$ 27,770.65
Limite inferior: R$ -28,484.74
Limite superior: R$ 82,597.85

Quantidade de outliers: 452
Percentual da base: 0.92%
Menor outlier: R$ 82,618.00
Maior outlier: R$ 127,262.02


Foram identificados 452 outliers, equivalentes a 0,92% da base. O primeiro quartil foi de R$ 13.171,24, o terceiro quartil foi de R$ 40.941,88 e o IQR foi de R$ 27.770,65. Com isso, foi definido um limite superior de R$ 82.597,85.

Não foram identificados outliers inferiores, pois conforme informado anteriormente, o limite inferior calculado foi negativo (R$ -28.484,74). Todos os 452 outliers são pedidos de alto valor, entre R$ 82.618,00 e R$ 127.262,02.

## 3.2 -  Impacto financeiro dos registros atípicos

Após a identificação dos outliers pela metodologia IQR, precisa ser avaliado se esses pedidos representam apenas uma ocorrência estatística pontual ou se possuem impacto relevante no faturamento da empresa.

Para isso, serão calculados:

- o faturamento total da base;
- o faturamento associado aos pedidos classificados como outliers;
- a participação percentual desses registros no faturamento total;
- a média dos pedidos após a exclusão temporária dos outliers.

A participação dos outliers no faturamento será calculada pela fórmula:

\[
\text{Participação dos outliers} =
\frac{\text{Faturamento dos outliers}}
{\text{Faturamento total}}
\times 100
\]

A média sem outliers será utilizada apenas como análise de sensibilidade, para comparar o comportamento dos pedidos típicos com a média original. Nenhum registro será removido da base.

In [20]:
# Soma o valor total de todos os pedidos da base.
faturamento_total = orders["total"].sum()

# Soma somente o valor dos pedidos identificados como outliers pelo IQR.
faturamento_outliers = outliers["total"].sum()

# Calcula o percentual do faturamento total concentrado nos outliers.
participacao_faturamento = (
    faturamento_outliers / faturamento_total
) * 100

# Remove temporariamente os índices presentes em outliers
# para calcular a média dos pedidos sem valores atípicos.
media_sem_outliers = orders.loc[
    ~orders.index.isin(outliers.index),
    "total"
].mean()

# Apresenta os resultados em formato monetário e percentual.
print(f"Faturamento total: R$ {faturamento_total:,.2f}")
print(f"Faturamento dos outliers: R$ {faturamento_outliers:,.2f}")
print(f"Participação no faturamento: {participacao_faturamento:.2f}%")
print(f"Média sem outliers: R$ {media_sem_outliers:,.2f}")

Faturamento total: R$ 1,406,487,201.80
Faturamento dos outliers: R$ 41,383,549.11
Participação no faturamento: 2.94%
Média sem outliers: R$ 28,119.80


Apesar de representarem uma parcela pequena da quantidade de pedidos, esses registros concentram **2,94%** do faturamento total. A média geral é R$ 28.704,99, enquanto a média sem os outliers é R$ 28.119,80, demonstrando que os pedidos extremos elevam a média.

## Diagnóstico inicial

Foi identificado que a tabela `orders` possui **48.998 registros** e **13 colunas**, cobrindo o período de **01/01/2020 a 31/12/2026**. A coluna `total` apresenta valores entre **R$ 32,62** e **R$ 127.262,02**, com média de **R$ 28.704,99** e mediana de **R$ 25.917,84**.

A análise pelo método IQR identificou **452 pedidos atípicos de alto valor**, equivalentes a **0,92% da base**. Esses pedidos possuem valor superior a **R$ 82.597,85** e representam **2,94% do faturamento total**. Portanto, embora sejam pouco frequentes, possuem impacto relevante nos indicadores financeiros e elevam a média dos pedidos. Não foram identificados outliers inferiores.

Os valores atípicos não devem ser removidos automaticamente, pois podem corresponder a compras legítimas de alto volume ou maior valor agregado. Recomenda-se validá-los com as regras de negócio antes de utilizá-los em metas, previsões ou comparações de desempenho. Para representar o comportamento típico de compra, a mediana e métricas segmentadas podem ser mais adequadas que a média isolada.

Não foram identificados valores nulos nas colunas financeiras ou de data. Contudo, `salesperson_id` possui **24.131 valores ausentes**, correspondentes a aproximadamente **49,25% dos pedidos**. Essa ausência pode estar relacionada a vendas de e-commerce, mas precisa ser confirmada com as regras da operação e com o relacionamento entre as demais tabelas.

### Conclui-se que **a tabela é adequada para análises exploratórias iniciais.** Entretanto, para análises que orientem decisões comerciais, previsões ou avaliação de vendedores, **recomenda-se validar os pedidos de alto valor e contextualizar os registros sem vendedor associado**.

## Questão 1.1 — Consulta SQL

Foi utilizada uma consulta agregada sobre a tabela `orders`, sem filtros e sem tratamento dos dados. Dessa forma, os indicadores representam a totalidade dos pedidos disponíveis na base.

As funções `MIN` e `MAX` foram aplicadas para identificar os limites temporais e financeiros. A função `AVG` foi utilizada para calcular o valor médio dos pedidos.

```sql
SELECT
    COUNT(*) AS total_linhas,
    MIN(created_at) AS data_minima,
    MAX(created_at) AS data_maxima,
    MIN(total) AS total_minimo,
    MAX(total) AS total_maximo,
    AVG(total) AS total_medio
FROM orders;
```

## Questão 1.2 - Validação
Qual é o valor médio registrado na coluna "total"?

### A leitura do resultado deve ser: a tabela possui 48.998 pedidos, com registros de 2020 a 2026, e valor médio de pedido de R$ 28.704,99.
